### ISSUE TITLE
WriteRecordsAsync - at runtime getting "This Operation is only valid on generic types."

### ISSUE THREAD

**Issue by MiblaCz** [JoshClose/CsvHelper#2081]
Writing objects into CSV throws "This Operation is only valid on generic types." runtime error.
Following code worked correctly until version 28.0.1, error occurs from 29.0.0 and even in 30.0.1.

```
IEnumerable<MyType> records = ...
await csv.WriteRecordsAsync<MyType>(records);
```

There is some problem with CsvHelper.CsvContext.AutoMap

I use only [Name("fieldname")] for properties of MyType, no custom classmaps.

![image](https://user-images.githubusercontent.com/118185301/201668991-fb8fc422-900c-4262-8ae6-6adc155b1bc1.png)

**Comment by JoshClose** [comment]
What does your type look like?

**Comment by MiblaCz** [comment]
I think, it is related with #2069

**Comment by MiblaCz** [comment]
@JoshClose please look at this at src/CsvHelper/TypeConversion/CollectionConverterFactory.cs line 120:

`		public bool Create(Type type, TypeConverterCache cache, out ITypeConverter typeConverter)
		{
		...
			var isGenericType = type.GetTypeInfo().IsGenericType;
			var genericTypeDefinition = type.GetGenericTypeDefinition();
			// this is called even if isGenericType == false

			if (isGenericType && genericTypeDefinition == typeof(Dictionary<,>))
			{
				typeConverter = new IDictionaryGenericConverter();
				return true;
			}
		...
		}
`

**Comment by MiblaCz** [comment]
Fixed in pull request #2086 more than month ago, but still not merged :-(

**Comment by Mark-Good** [comment]
This seems like a really strange and nasty bug that testing should have caught.  

I rolled back to 28.0.1.  🤷‍♂️
